{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0280bdb3-2805-411c-bb0c-ea4b31d85018",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Versión 1.1.0 de Coronalyzer. Programa desarrollado para analizar las secuencias de las diferentes variantes del Sars-Cov2. \n",
    "Tipo de análisis: Análisis de Secuencia, Análisis de mutaciones, Análisis de Similitud entre secuencias.\n",
    "Función añadida: Descarga de archvivos PDB, Visualización \n",
    "Funciones extra: Descarga de Fasta y Fasta cds, preparación de fastas para clustl (Align Tree)\n",
    "Autor: Pedro Manuel López Zarzuela \n",
    "Fecha: 03/12/2021\n",
    "'''\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import os\n",
    "import math\n",
    "from tabulate import tabulate\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "pd.plotting.register_matplotlib_converters()\n",
    "import seaborn as sns\n",
    "import Bio\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio import AlignIO\n",
    "from Bio import Entrez\n",
    "from Bio import ExPASy\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio.PDB import MMCIFParser\n",
    "#from Bio import Phylo\n",
    "from tkinter import *\n",
    "import tkinter as tk\n",
    "from prody import *\n",
    "from numpy import *\n",
    "from matplotlib.pyplot import *\n",
    "import matplotlib.image as mpimg\n",
    "import pylab\n",
    "import pandas as pd\n",
    "import nglview as nv\n",
    "\n",
    "text1=\"\"\n",
    "text2=\"\"\n",
    "text3=\"\"\n",
    "text4=\"\"\n",
    "text5=\"\"\n",
    "text6=\"\"\n",
    "text7=\"\"\n",
    "text8=\"\"\n",
    "\n",
    "'''Función do_nothing() para ejectutar el menú de tkinter sin llamar a las funciones del programa y que pueda ejecutarse'''\n",
    "\n",
    "def donothing():\n",
    "   x = 0\n",
    "\n",
    "'''Función limpiarCampos(), función que limpia los campos de las entradas (Entry) de nuestro menu contextual de Tkinter'''\n",
    "\n",
    "def limpiarCampos():\n",
    "    \n",
    "    miNombre1.set(\"\")\n",
    "    miNombre2.set(\"\")\n",
    "    miNombre3.set(\"\")\n",
    "    miNombre4.set(\"\")\n",
    "    miNombre5.set(\"\")\n",
    "    \n",
    "'''Función descargar_fasta(), como su propio nombre indica nos descarga los fastas introduciendo el codigo de del NCBI\n",
    "descarga los fasta y los fasta_cds'''\n",
    "\n",
    "def descargar_fasta():\n",
    "    global text4\n",
    "    text4=nombreID4.get()\n",
    "    print(text4)\n",
    "    \n",
    "    Entrez.email = \"8tato8@gmail.com\"\n",
    "    hd1 = Entrez.efetch(db=\"nucleotide\",id=[text4],rettype='fasta')\n",
    "    seq = SeqIO.read(hd1,'fasta')\n",
    "    fw = open(text4,'w')\n",
    "    SeqIO.write(seq,fw,'fasta')\n",
    "    \n",
    "    \n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=[text4], rettype=\"fasta_cds_na\", retmode=\"text\")\n",
    "    seq=handle.read()\n",
    "    text_file = open(text4+'.txt', \"w\")\n",
    "    n = text_file.write(seq)\n",
    "    \n",
    "    text_file.close()\n",
    "    \n",
    "    fw.close()\n",
    "    os.getcwd()\n",
    "    \n",
    "    return(text4)\n",
    "\n",
    "'''\n",
    "Función desarrollada para aliniear los diferentes fastas de las variantes y subirlo al Clustl para realizar un phylogenetic Tree\n",
    "def unir_fastas():\n",
    "    #import the sequences we will use. These are 16s sequences from GenBank\n",
    "    #example: https://www.ncbi.nlm.nih.gov/nuccore/FJ039971.1?report=genbank \n",
    "    t1 = Bio.SeqIO.read(\"229E.fasta\", \"fasta\")\n",
    "    t2 = Bio.SeqIO.read(\"HK13.fasta\", \"fasta\")\n",
    "    t3 = Bio.SeqIO.read(\"Wuhan.fasta\", \"fasta\")\n",
    "    t4 = Bio.SeqIO.read(\"HKU1.fasta\", \"fasta\")\n",
    "    t5 = SeqIO.read(\"Alpha.fasta\", \"fasta\")\n",
    "    t6 = SeqIO.read(\"Beta.fasta\", \"fasta\")\n",
    "    t7 = SeqIO.read(\"Delta.fasta\", \"fasta\")\n",
    "    t8 = SeqIO.read(\"lambda.fasta\", \"fasta\")\n",
    "    t9 = SeqIO.read(\"Gamma.fasta\", \"fasta\")\n",
    "    #rename each of the sequences \n",
    "    #this step is not required, it will just make the tree easier to understand \n",
    "    print(t1.description)\n",
    "    print(t2.description)\n",
    "    print(t3.description)\n",
    "    print(t4.description)\n",
    "\n",
    "    t1.id = '229E'\n",
    "    t2.id = 'HK13'\n",
    "    t3.id = 'Wuhan'\n",
    "    t4.id = 'HKU1'\n",
    "    t5.id = 'Alpha'\n",
    "    t6.id = 'Beta'\n",
    "    t7.id = 'Delta'\n",
    "    t8.id = 'lambda'\n",
    "    t9.id = 'Gamma'\n",
    "\n",
    "    # Combine all of the individual sequences into a new file \n",
    "    variantes = SeqIO.write([t1,t2,t3,t4,t5,t6,t7,t8,t9], \"variantes.fasta\", \"fasta\")\n",
    "'''\n",
    "\n",
    "'''Función descargar_PDB, como su propio nombre indica descarga archivos del PDB introduciendo el codigo de 4 letras'''\n",
    "\n",
    "def descargar_PDB():\n",
    "    global text5\n",
    "    text5=nombreID5.get()\n",
    "    pdbl=PDBList()\n",
    "    pdbl1=[text5]\n",
    "    for i in pdbl1:\n",
    "        pdbl.download_pdb_files(pdbl1, pdir='PDB', file_format='mmCif')\n",
    "    return(text5)\n",
    "\n",
    "'''Función Analizar_Seq(), hace un analisis de la secuencia obtenida a partir de un archivo .fasta: \n",
    "Imprime secuencia, la traduce a mRNA, transcribe a proteina, reliza el splicin, la RC y ORF'''\n",
    "\n",
    "def Analizar_Seq():\n",
    "    global text1\n",
    "\n",
    "    text1=nombreID1.get()\n",
    "\n",
    "    print(text1)\n",
    "    \n",
    "    coronavirus = open(text1, \"r\")\n",
    " \n",
    "    print('HEADER:',coronavirus.readline())\n",
    "    coronavirus = coronavirus.readlines()\n",
    "    COVID_seq = ''\n",
    "    for line in coronavirus:\n",
    "        line = line.strip()\n",
    "        COVID_seq += line\n",
    "    print(COVID_seq[0:1000])\n",
    "    \n",
    "    def basic_properties(DNAseq):\n",
    "        total_base = len(DNAseq)\n",
    "        num_Adenine = DNAseq.count('A')\n",
    "        num_Guanine = DNAseq.count('G')\n",
    "        num_Thymine = DNAseq.count('T')\n",
    "        num_Cytosine = DNAseq.count('C')\n",
    "        \n",
    "        if total_base != num_Adenine + num_Guanine + num_Thymine + num_Cytosine:\n",
    "            print('Something is not right')\n",
    "        else : pass\n",
    "        \n",
    "        A_percent = num_Adenine / total_base\n",
    "        G_percent = num_Guanine / total_base\n",
    "        T_percent = num_Thymine / total_base\n",
    "        C_percent = num_Cytosine / total_base\n",
    "        \n",
    "        #visualization\n",
    "        x = np.arange(4)\n",
    "        bases = ['Adenine', 'Guanine', 'Thymine' ,'Cytosine']\n",
    "        values = [num_Adenine, num_Guanine, num_Thymine, num_Cytosine]\n",
    "        plt.bar(x,values)\n",
    "        plt.xticks(x, bases)\n",
    "        plt.show()\n",
    "        table = [['total base',total_base,'Percentage',str('100%')],\n",
    "                 ['Adenine:',num_Adenine, 'Percentage:',str(round(A_percent*100,2))+'%'],\n",
    "                ['Guanine:',num_Guanine, 'Percentage:',str(round(G_percent*100,2))+'%'],\n",
    "                 ['Thynime:',num_Thymine, 'Percentage:',str(round(T_percent*100,2))+'%'],\n",
    "                 ['Cytosine:',num_Cytosine, 'Percentage:',str(round(C_percent*100,2))+'%']]\n",
    "        print(tabulate(table))\n",
    "        print('GC content:', round((((num_Guanine + num_Cytosine) / total_base)*100),2),'%')\n",
    "        \n",
    "    basic_properties(COVID_seq)\n",
    "    \n",
    "    def transcription(DNAseq):\n",
    "        mRNAseq = DNAseq.replace('T','U')\n",
    "        print(len(mRNAseq))\n",
    "        return mRNAseq\n",
    "    COVID_mRNA = transcription(COVID_seq)\n",
    "    print(\"----------------mRNA sequence---------------------\")\n",
    "    print(COVID_mRNA[0:1000])\n",
    "    print(\"---------------Protein Sequence-------------------\")\n",
    "    \n",
    "    def translate(seq):\n",
    "        table = {\n",
    "            'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "            'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "            'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "            'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n",
    "            'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "            'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "            'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "            'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "            'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "            'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "            'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "            'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "            'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "            'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "            'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "            'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',}\n",
    "        protein =\"\"\n",
    "        for i in range(0, len(seq), 3):\n",
    "            codon = seq[i:i + 3]\n",
    "            if codon in table:\n",
    "                protein+= table[codon]\n",
    "        return protein\n",
    "    \n",
    "    protein_seq = translate(COVID_seq)\n",
    "    print(protein_seq[0:1000])\n",
    "    \n",
    "    protein_seq = translate(COVID_seq)\n",
    "    protein_seq[0:1000]\n",
    "    \n",
    "    def visualization(protein_seq):\n",
    "        # composición de los Aminoácidos\n",
    "        plt.figure(figsize=(35,50))\n",
    "        x = np.arange(22)\n",
    "        AA = ['Arginine', 'Histidie','Lysine',\n",
    "              'Aspartic Acid','Glutamic Acid',\n",
    "              'Serine','Threonine','Asparagine','Glutamine',\n",
    "              'Cysteine','Selenocysteine','Glycine','Prolien',\n",
    "              'Alanine','Valine','Isoleucine','Leucine',\n",
    "              'Methionine','Phenylalanine','Tyrosine','Tryptophan',\n",
    "              'Stop Codon']\n",
    "        values = [protein_seq.count('R'),protein_seq.count('H'),protein_seq.count('K'),\n",
    "                  protein_seq.count('D'),protein_seq.count('E'),\n",
    "                  protein_seq.count('S'),protein_seq.count('T'),protein_seq.count('N'),protein_seq.count('Q'),\n",
    "                  protein_seq.count('C'),protein_seq.count('U'),protein_seq.count('G'),protein_seq.count('P'),\n",
    "                  protein_seq.count('A'),protein_seq.count('V'),protein_seq.count('I'),protein_seq.count('L'),\n",
    "                  protein_seq.count('M'),protein_seq.count('F'),protein_seq.count('Y'),protein_seq.count('W'),\n",
    "                  protein_seq.count('_')]\n",
    "        plt.subplot(2,2,1)\n",
    "        plt.rc('font',size = 20)\n",
    "        plt.barh(AA,values,height=0.6)\n",
    "        plt.title('AA in general')\n",
    "        \n",
    "        # Visualización de grupos segun su carga \n",
    "        x = np.arange(4)\n",
    "        Electric =  protein_seq.count('R')+protein_seq.count('H')+protein_seq.count('K')+protein_seq.count('D')+protein_seq.count('E')\n",
    "        Uncharged = protein_seq.count('S')+protein_seq.count('T')+protein_seq.count('N')+protein_seq.count('Q')\n",
    "        Special =   protein_seq.count('C')+protein_seq.count('U')+protein_seq.count('G')+protein_seq.count('P')\n",
    "        Hydrophobic = protein_seq.count('A')+protein_seq.count('V')+protein_seq.count('I')+protein_seq.count('L')+protein_seq.count('M')+protein_seq.count('F')+protein_seq.count('Y')+protein_seq.count('W')\n",
    "        \n",
    "        plt.subplot(2,2,2)\n",
    "        types = ['Elecrically charged','Polar uncharged',\n",
    "                 'Special case', 'Hydrophobic Side Chain']\n",
    "        values = [Electric,Uncharged,Special,Hydrophobic]\n",
    "        plt.barh(types, values, height = 0.6)\n",
    "        plt.title('AA in groups')\n",
    "        \n",
    "        # Visualización de los aminoacidos positivos y negativos en grupos de cadena laterales hidrofóbicas\n",
    "        plt.subplot(2,2,3)\n",
    "        x = np.arange(2)\n",
    "        positive = protein_seq.count('R')+protein_seq.count('H')+protein_seq.count('K')\n",
    "        negative = protein_seq.count('D')+protein_seq.count('E')\n",
    "        types = ['Positive','Negative']\n",
    "        values = [positive, negative]\n",
    "        plt.barh(types,values, height=0.6)\n",
    "        plt.title('Charge diff in Electrically charged side chain')\n",
    "        plt.show()\n",
    "        \n",
    "        # Visualizacion de la abundancia de Aminoacidos segun su estructura secundaria en la cadena proteica \n",
    "        alpha_helix = protein_seq.count('A')+protein_seq.count('C')+protein_seq.count('L')+protein_seq.count('M')+protein_seq.count('E')+protein_seq.count('Q')+protein_seq.count('H')+protein_seq.count('K')\n",
    "        beta_sheet =protein_seq.count('V')+protein_seq.count('I')+protein_seq.count('F')+protein_seq.count('Y')+protein_seq.count('W')+protein_seq.count('T')\n",
    "        turn = protein_seq.count('G')+protein_seq.count('S')+protein_seq.count('D')+protein_seq.count('N')+protein_seq.count('P')\n",
    "        x = np.arange(3)\n",
    "        plt.subplot(2,2,4)\n",
    "        types = ['Alpha helix','beta sheet','Turn']\n",
    "        values = [alpha_helix, beta_sheet, turn]\n",
    "        plt.barh(types,values, height=0.6)\n",
    "        plt.title('AA residues in secondary structure')\n",
    "        plt.show()\n",
    "        \n",
    "    visualization(protein_seq)\n",
    "    \n",
    "    print(\"------------------------------------------------------Protein Splicing-------------------------------------------------------\")\n",
    "    def proteinseq_splic(protein_sequence):\n",
    "        protein_group = protein_sequence.split('_')\n",
    "        return protein_group\n",
    "    protein_list = proteinseq_splic(protein_seq)\n",
    "    print(proteinseq_splic(protein_seq)[0:100])\n",
    "    print('Number of proteins:',len(protein_list))\n",
    "    \n",
    "    print(\"-----------------------------------------------------Functional Protein-------------------------------------------------------\")\n",
    "    def Functional_protein(protein_list):\n",
    "        funct_group = []\n",
    "        for protein in protein_list:\n",
    "            if len(protein) > 20:\n",
    "                funct_group.append(protein)\n",
    "        return funct_group\n",
    "    function_protein = Functional_protein(protein_list)\n",
    "    print(function_protein[0:52])\n",
    "    print('Number of Functional proteins:',len(function_protein))\n",
    "    \n",
    "    dna_ORF = []\n",
    "    \n",
    "    def rf(seq):\n",
    "        #getting all reading frames\n",
    "        i=0\n",
    "        while i+2<len(seq):\n",
    "            if seq[i:i+3]=='ATG':\n",
    "                j=i+3\n",
    "                while j+2<len(seq):\n",
    "                    if seq[j:j+3] in ['TGA', 'TAA', 'TAG']:\n",
    "                        dna_ORF.append(seq[i:j+3])\n",
    "                    j=j+3\n",
    "            i=i+3\n",
    "        return dna_ORF\n",
    "    \n",
    "    def RC(seq):\n",
    "        #reverse compliment\n",
    "        complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "        bases = list(seq) \n",
    "        bases = reversed([complement.get(base,base) for base in bases])\n",
    "        bases = ''.join(bases)\n",
    "        return bases\n",
    "       \n",
    "    def ORF(seq):\n",
    "        #open reading frame\n",
    "        #num=0\n",
    "        seqRC=RC(seq)\n",
    "        for i in [0,1,2]:\n",
    "            rf(seq[i:])\n",
    "            rf(seqRC[i:])\n",
    "            \n",
    "    RC(COVID_seq)[0:1000]\n",
    "    print(\"--------------------------------------------------DNA OPEN READING FRAME------------------------------------------------------\")\n",
    "    ORF(COVID_seq)\n",
    "    record = SeqIO.read(text1, \"fasta\")\n",
    "    table = 11\n",
    "    min_pro_len = 100\n",
    "    for strand, nuc in [(+1, record.seq), (-1, record.seq.reverse_complement())]:\n",
    "         for frame in range(3):\n",
    "             length = 3 * ((len(record)-frame) // 3) #Multiple of three\n",
    "             for pro in nuc[frame:frame+length].translate(table).split(\"*\"):\n",
    "                 if len(pro) >= min_pro_len:\n",
    "                     print(\"%s...%s - length %i, strand %i, frame %i\" \\\n",
    "                           % (pro[:30], pro[-3:], len(pro), strand, frame))\n",
    " \n",
    "    return (text1,text2,text3)\n",
    "    \n",
    "def Analizar_Mut():\n",
    "    global text1\n",
    "    global text2\n",
    "    text1=nombreID1.get()\n",
    "    text2=nombreID2.get()\n",
    "    print(text1,text2)\n",
    "  \n",
    "    # Este diccionario se usa para restablecer el tamaño de la numpy array por cada uno de los genes. \n",
    "    # La numpy array del gen  ORF1ab se establecera con rows=115 y cols=115.\n",
    "    # El segundo elemento de la lista hace referencia a el numero 'n' dummy Nucleotides que seran añadidos al final\n",
    "    # de cada gen para hacer las secuencias compactas.\n",
    "    numpy_image_dict = {'gene=ORF1ab':[(115,115),7],\n",
    "                        'gene=S':[(62,62),22],\n",
    "                        'gene=ORF3a':[(28,30),12],\n",
    "                        'gene=E':[(15,16),12], \n",
    "                        'gene=M':[(26,27),33],\n",
    "                        'gene=ORF6':[(14,14),10],    \n",
    "                        'gene=ORF7a':[(19,20),14],\n",
    "                        'gene=ORF7b':[(12,12),12],\n",
    "                        'gene=ORF8':[(19,20),14],\n",
    "                        'gene=N':[(36,36),36],\n",
    "                        'gene=ORF10':[(11,11),4]}\n",
    "    \n",
    "    \n",
    "    # Este diccionario tiene los codones para transofrar mRNA a aminoacidos.\n",
    "    amacid_dict = {'UUU':('F','PHE','Phenylalanine'),'UUC':('F','PHE','Phenylalanine'),\n",
    "                   'UUA':('L','LEU','Leucine'),'UUG':('L','LEU','Leucine'),\n",
    "                   'UCU':('S','SER','Serine'),'UCC':('S','SER','Serine'),\n",
    "                   'UCA':('S','SER','Serine'),'UCG':('S','SER','Serine'), \n",
    "                   'UAU':('Y','TYR','Tyrosine'),'UAC':('Y','TYR','Tyrosine'),\n",
    "                   'UAA':('STOP'),'UAG':('STOP'),\n",
    "                   'UGU':('C','CYS','Cysteine'),'UGC':('C','CYS','Cysteine'),\n",
    "                   'UGA':('STOP'),'UGG':('W','TRP','Tryptophan'),\n",
    "                   'CUU':('L','LEU','Leucine'),'CUC':('L','LEU','Leucine'),\n",
    "                   'CUA':('L','LEU','Leucine'),'CUG':('L','LEU','Leucine'),\n",
    "                   'CCU':('P','PRO','Proline'),'CCC':('P','PRO','Proline'),\n",
    "                   'CCA':('P','PRO','Proline'),'CCG':('P','PRO','Proline'),\n",
    "                   'CAU':('H','HIS','Histidine'),'CAC':('H','HIS','Histidine'),\n",
    "                   'CAA':('Q','GLU','Glutamine'),'CAG':('Q','GLU','Glutamine'),\n",
    "                   'CGU':('R','ARG','Arginine'),'CGA':('R','ARG','Arginine'),\n",
    "                   'CGG':('R','ARG','Arginine'),'CGC':('R','ARG','Arginine'),\n",
    "                   'AUU':('I','ILE','Isoleucine'),'AUC':('I','ILE','Isoleucine'),\n",
    "                   'AUA':('I','ILE','Isoleucine'),'AUG':('M','MET','Methionine'),\n",
    "                   'ACU':('T','THR','Threonine'),'ACC':('T','THR','Threonine'),\n",
    "                   'ACA':('T','THR','Threonine'),'ACG':('T','THR','Threonine'),\n",
    "                   'AAU':('N','ASN','Asparagine'),'AAC':('N','ASN','Asparagine'),\n",
    "                   'AAA':('K','LYS','Lysine'),'AAG':('K','LYS','Lysine'),\n",
    "                   'AGU':('S','SER','Serine'),'AGC':('S','SER','Serine'),\n",
    "                   'AGG':('R','ARG','Arginine'),'AGA':('R','ARG','Arginine'),\n",
    "                   'GUU':('V','VAL','Valine'),'GUC':('V','VAL','Valine'),\n",
    "                   'GUA':('V','VAL','Valine'),'GUG':('V','VAL','Valine'),\n",
    "                   'GCU':('A','ALA','Alanine'),'GCC':('A','ALA','Alanine'),\n",
    "                   'GCA':('A','ALA','Alanine'),'GCG':('A','ALA','Alanine'),\n",
    "                   'GAU':('D','ASP','Aspartate'),'GAC':('D','ASP','Aspartate'),\n",
    "                   'GAA':('E','GLU','Glutamate'),'GAG':('E','GLU','Glutamate'),\n",
    "                   'GGU':('G','GLY','Glycine'),'GGC':('G','GLY','Glycine'),\n",
    "                   'GGA':('G','GLY','Glycine'),'GGG':('G','GLY','Glycine')\n",
    "    }\n",
    "    #Este es el modulo principal .\n",
    "    class dna:\n",
    "    \n",
    "        # metodo cosntructor\n",
    "        def __init__(self,dna_seq):\n",
    "            dna_seq = dna_seq.upper() # Convirte bases en mayusculas \n",
    "            for seq in dna_seq:\n",
    "                # Valida las bases. Si la secuencia no es valida salta un error\n",
    "                if seq not in ['A','T','G','C',' ','N']: \n",
    "                    error = 'Wrong DNA Sequence {}!!'.format(seq)\n",
    "                    raise ValueError(error)\n",
    "            # Quita todos los caracteres vacios de la secuencia.\n",
    "            dna_seq = dna_seq.replace(' ','') \n",
    "            self.dir_3_5=dna_seq\n",
    "            self.dir_5_3=self.dir_5_3_strand()\n",
    "            self.mRna = None\n",
    "            self.amino_acid = None\n",
    "            self.num_array = None\n",
    "            self.nucl_len = len(dna_seq)\n",
    "    \n",
    "        def __repr__(self):\n",
    "            return \"DNA has {} nucleotide and they are {} :\".format(self.nucl_len,self.dir_3_5)\n",
    "    \n",
    "        def __eq__(self, other):\n",
    "            if other is None:\n",
    "                return False\n",
    "            return self.seq == other.seq\n",
    "    \n",
    "        def transcription(self):\n",
    "            # Este metodo traduce DNA a mRNA\n",
    "            trans=''\n",
    "            for nuc in self.dir_5_3:\n",
    "                if nuc == 'A':\n",
    "                    trans += 'U'\n",
    "                if nuc == 'T':\n",
    "                    trans += 'A'\n",
    "                if nuc == 'C':\n",
    "                    trans += 'G'\n",
    "                if nuc == 'G':\n",
    "                    trans += 'C'\n",
    "                if nuc == 'N':\n",
    "                    trans += 'N'\n",
    "            self.mRna = trans\n",
    "            return self.mRna\n",
    "    \n",
    "        def translation(self):\n",
    "            # Este método usa el DNa traducido a mRNa para transcribirlo a su correspondiente secuencia de AA\n",
    "            begin = 'No'\n",
    "            ac = ''\n",
    "            for i in range(0,len(self.mRna)-3,3):\n",
    "                if self.mRna[i:3] == 'AUG':\n",
    "                    begin = 'Yes'\n",
    "                if self.mRna[i:3] in ('UAA','UAG','UGA'):\n",
    "                    being = 'No'\n",
    "                if begin == 'Yes':\n",
    "                    ac+= amacid_dict[self.mRna[i:3+i]][0]\n",
    "            self.amino_acid = ac\n",
    "            return self.amino_acid\n",
    "    \n",
    "        def dir_5_3_strand(self):\n",
    "            dir_5_3 = ''\n",
    "            # TEste metodo pasa la secuencia 3'-5' a 5'-3'.\n",
    "            for nuc in self.dir_3_5:\n",
    "                if nuc == 'A':\n",
    "                    dir_5_3 += 'T'\n",
    "                if nuc == 'T':\n",
    "                    dir_5_3 += 'A'\n",
    "                if nuc == 'C':\n",
    "                    dir_5_3 += 'G'\n",
    "                if nuc == 'G':\n",
    "                    dir_5_3 += 'C'\n",
    "                if nuc == 'N':\n",
    "                    dir_5_3 += 'N'\n",
    "            return dir_5_3\n",
    "    \n",
    "        def numpfy(self):\n",
    "            # Este metodo comnvierte el Dna en un numpy array.\n",
    "            # Cada nucleotido de la secuencia se traduce a un numero que vemos a continuación \n",
    "            # Esto puede ser usado para analisis o comparación, como el One-Hot-Encoding.\n",
    "            arr = ''\n",
    "            for i in self.dir_3_5:\n",
    "                if i == 'A':\n",
    "                    arr += '0 '\n",
    "                if i == 'T':\n",
    "                    arr += '255 '\n",
    "                if i == 'C':\n",
    "                    arr += '100 '\n",
    "                if i == 'G':\n",
    "                    arr += '200 '\n",
    "                if i == 'N':\n",
    "                    arr += '75 '   \n",
    "            arr_np = np.fromstring(arr,dtype=np.uint8,sep=' ')        \n",
    "            self.num_array = arr_np\n",
    "            return self.num_array\n",
    "        \n",
    "        \n",
    "        \n",
    "    def read_dna_seq(file_name):\n",
    "        global fil\n",
    "        # Este metodo lee la secuencia fasta descargada del NCBI y crea un diccionario con ella.\n",
    "        try:\n",
    "            fil = open(file_name,'r')\n",
    "            fil_list = fil.readlines()\n",
    "            fil.close\n",
    "        except: FileNotFoundError()\n",
    "        \n",
    "        genome = {}\n",
    "        gene_name = ''\n",
    "        protein_name = ''\n",
    "        gene_seq = ''\n",
    "        for i in fil_list:\n",
    "            if i[0] == '>':\n",
    "                # Leemos cada linea del archivo y creamos un diccionario con la siguiente información\n",
    "                # gene. {<'gene_name-1'>:[<protein_name>,nucleotide sequence],\n",
    "                #        <'gene_name-2'>:[<protein_name>,nucleotide sequence],\n",
    "                #        <'gene_name-2'>:[<protein_name>,nucleotide sequence]}\n",
    "                if list(genome.keys()) != []:\n",
    "                    gene_seq = gene_seq.replace('\\n','')\n",
    "                    genome[gene_name].append(gene_seq)\n",
    "                gene_seq = ''\n",
    "                g_st = i.find('[gene=')\n",
    "                g_end = i[g_st:].find(']')\n",
    "                p_st = i.find('[protein=')\n",
    "                p_end = i[p_st:].find(']') \n",
    "    \n",
    "                if g_st > 0 and g_end > 0:\n",
    "                    gene_name = i[g_st+1:g_st+g_end]\n",
    "                    genome[gene_name] = []\n",
    "                \n",
    "                if p_st > 0 and p_end > 0:\n",
    "                    protein_name = i[p_st+1:p_st+p_end]\n",
    "                    genome[gene_name].append(protein_name)\n",
    "            else:\n",
    "                gene_seq += i\n",
    "        gene_seq = gene_seq.replace('\\n','')\n",
    "        genome[gene_name].append(gene_seq)    \n",
    "        return genome\n",
    "    \n",
    "    def gene_mod(genome):\n",
    "        # Substitución por dummy nucleotide 'N' para compactar la numpy array.\n",
    "        genome_keys = list(genome.keys())\n",
    "        for k in genome_keys:\n",
    "            if len(numpy_image_dict[k]) > 1:\n",
    "                N = numpy_image_dict[k][1]\n",
    "                seq = add_N(N,genome[k][1])\n",
    "                genome[k][1] = seq\n",
    "        return genome\n",
    "    \n",
    "    def add_N(n,seq):\n",
    "        # Esta funcion la usa gene_mod para añadir dummy 'N'.\n",
    "        for i in range(0,n):\n",
    "            seq += 'N'\n",
    "        return seq\n",
    "    \n",
    "    # Lee la secuencia 1 de DNA descargada anteriormente del NCBI.\n",
    "    dict_seq_1 = read_dna_seq(text1)\n",
    "    # Modifica la secuencia con dummy 'N' nucleotide.\n",
    "    dict_seq_1 = gene_mod(dict_seq_1)\n",
    "    \n",
    "    # Lee la secuencia 2 de DNA descargada anteriormente del NCBI.\n",
    "    dict_seq_2 = read_dna_seq(text2)\n",
    "    # Modifica la secuencia con dummy 'N' nucleotide.\n",
    "    dict_seq_2 = gene_mod(dict_seq_2)\n",
    "    \n",
    "    # Creamos plots con Matplotlib para cada gen. \n",
    "    f,ax = plt.subplots(nrows=11,ncols=3,figsize=(25,30))\n",
    "    gene_name = list(numpy_image_dict.keys())\n",
    "    row = 0\n",
    "    col = 0\n",
    "    mut_dict={}\n",
    "    for i in gene_name:\n",
    "        G = i[5:]\n",
    "        # Loopeamos a traves de cada gen de la secuencia de Cornona Virus.\n",
    "        gene_us = dna(dict_seq_1['gene='+G][1])\n",
    "        # Invocamos el metodo de transcription \n",
    "        gene_us.transcription()\n",
    "        # Invocamos el metodo que convierte gene la secuencia de genes en un numpy array.\n",
    "        numpfy_usa = gene_us.numpfy()\n",
    "        # Reestablecemos las dimensiones del numpy array.\n",
    "        numpfy_usa = numpfy_usa.reshape(numpy_image_dict['gene='+G][0])\n",
    "        # Hacemos plot del array.\n",
    "        ax[row][col].pcolor(numpfy_usa)\n",
    "        ax[row][col].set_title(G+' Gene - USA')\n",
    "        col+=1\n",
    "        gene_china = dna(dict_seq_2['gene='+G][1])\n",
    "        # Invocamos el metodo de transcription\n",
    "        gene_china.transcription()\n",
    "        # Invocamos el metodo que convierte gene la secuencia de genes en un numpy array.\n",
    "        numpfy_china = gene_china.numpfy()\n",
    "        # Reestablecemos las dimensiones del numpy array.\n",
    "        numpfy_china = numpfy_china.reshape(numpy_image_dict['gene='+G][0])\n",
    "        # Hacemos el plot de la array\n",
    "        ax[row][col].pcolor(numpfy_china)\n",
    "        ax[row][col].set_title(G+' Gene - CHINA')\n",
    "        col+=1\n",
    "    \n",
    "        # Para encontrar las mutaciones lo que hacemos es restar la nueva secuencia a la  secuencia base.\n",
    "        # Chinese sequence is the base sequence and the USA sequence is a newer sequence.\n",
    "        mut = numpfy_china - numpfy_usa\n",
    "        if mut.any():\n",
    "            # Here we are looking for a non zero value in the mutated numpy array (result of the subtracting the 2 numpy arrays).\n",
    "            # La presencia de un valor diferente de zero nos indica que en esa localizacion existe una mutación para ese nucleotido . \n",
    "            #{'<Gene_Name-1>': [[<value_of_base_seq>, <value_of_newer_seq>, <value_in_mutated_numpy>, (x_value,y_value)]], '<Gene_Name-2>': [[<value_of_base_seq>, <value_of_newer_seq>, <value_in_mutated_numpy>, (x_value,y_value)]]}\n",
    "            mut_nec = np.nonzero(mut)\n",
    "            x=mut_nec[0]\n",
    "            y=mut_nec[1]\n",
    "            l=0\n",
    "            mut_dict[G]=[]\n",
    "            for i in x:\n",
    "                us_base = numpfy_usa[i][y[l]]\n",
    "                ch_base = numpfy_china[i][y[l]]\n",
    "                mut_base = mut[i][y[l]]\n",
    "                info_list = [ch_base,us_base,mut_base,(i,y[l])]\n",
    "                mut_dict[G].append(info_list)\n",
    "                print(\"Mutated DNA Base {} in China and Base {} in USA at position {} For the Gene {}\".format(ch_base,us_base,(i,y[l]),G))\n",
    "                l+= 1\n",
    "        # Poniendole título al plot\n",
    "        ax[row][col].pcolor(mut)\n",
    "        ax[row][col].set_title(G+' Gene - Mutataion')\n",
    "        row+= 1\n",
    "        col=0\n",
    "    \n",
    "    f.tight_layout()\n",
    "    # Guardamos el matplotlib subplot como un jpg.\n",
    "    f.savefig('Sars_Cov-2_Gene_Mutation.jpg')\n",
    "    img = mpimg.imread('Sars_Cov-2_Gene_Mutation.jpg')\n",
    "    imgplot = plt.imshow(img)\n",
    "    #plt.show()\n",
    "    return (text1,text2,text3)\n",
    "\n",
    "def Analizar_Sim():\n",
    "    global text1\n",
    "    global text2\n",
    "    \n",
    "    text1=nombreID1.get()\n",
    "    text2=nombreID2.get()\n",
    "    print(text1,text2)\n",
    "  \n",
    "    Seq1 = SeqIO.read(text1, 'fasta')\n",
    "    Seq2 = SeqIO.read(text2, 'fasta') \n",
    "    \n",
    "    Seq2_Seq1 = pairwise2.align.globalxx(Seq2.seq, Seq1.seq, one_alignment_only=True, score_only=True)\n",
    "    print('Seq2/Seq1 Similarity (%):', Seq2_Seq1 / len(Seq1.seq) * 100)\n",
    "    \n",
    "      # Printeamos los plots con la data\n",
    "    X = ['Seq2/Seq1']\n",
    "    Y = [Seq2_Seq1/len(Seq1.seq)*100]\n",
    "    plt.title('Sequence identity (%)')\n",
    "    plt.bar(X,Y,color=(0.2, 0.4, 0.6, 0.6))\n",
    "    plt.show()\n",
    "    return (text1,text2)\n",
    "    \n",
    "def Analizar_Mult():\n",
    "    global text1\n",
    "    global text2\n",
    "    global text3\n",
    "    \n",
    "    text1=nombreID1.get()\n",
    "    text2=nombreID2.get()\n",
    "    text3=nombreID3.get()\n",
    "    print(text1,text2,text3)\n",
    "  \n",
    "    Seq1 = SeqIO.read(text1, 'fasta')\n",
    "    Seq2 = SeqIO.read(text2, 'fasta')\n",
    "    Seq3 = SeqIO.read(text3, 'fasta')\n",
    "    \n",
    "    Seq1_Seq3 = pairwise2.align.globalxx(Seq1.seq, Seq3.seq, one_alignment_only=True, score_only=True)\n",
    "    print('Seq1/Seq3 Similarity (%):', Seq1_Seq3 / len(Seq1.seq) * 100)\n",
    "    \n",
    "    Seq2_Seq3 = pairwise2.align.globalxx(Seq2.seq, Seq3.seq, one_alignment_only=True, score_only=True)\n",
    "    print('Seq2/Seq3 Similarity (%):', Seq2_Seq3 / len(Seq2.seq) * 100)\n",
    "    \n",
    "    Seq2_Seq1 = pairwise2.align.globalxx(Seq2.seq, Seq1.seq, one_alignment_only=True, score_only=True)\n",
    "    print('Seq2/Seq1 Similarity (%):', Seq2_Seq1 / len(Seq1.seq) * 100)\n",
    "    \n",
    "    \n",
    "    # Printeamos la data\n",
    "    X = ['Seq1/Seq3', 'Seq2/Seq3', 'Seq2/Seq1']\n",
    "    Y = [Seq1_Seq3/ len(Seq1.seq) * 100, Seq2_Seq3/ len(Seq2.seq)*100, Seq2_Seq1/len(Seq1.seq)*100]\n",
    "    plt.title('Sequence identity (%)')\n",
    "    plt.bar(X,Y,color=(0.2, 0.4, 0.6, 0.6))\n",
    "    plt.show()\n",
    "    \n",
    "    return (text1,text2,text3)\n",
    "\n",
    "def protein_aln():\n",
    "    \n",
    "    global text1\n",
    "    global text2\n",
    "    \n",
    "    text1=nombreID1.get()\n",
    "    text2=nombreID2.get()\n",
    "    print(text1,text2)\n",
    "\n",
    "    with ExPASy.get_sprot_raw(text1) as handle:\n",
    "        seq_record1 = SeqIO.read(handle, \"swiss\")\n",
    "    print(seq_record1.id)\n",
    "    print(seq_record1.name)\n",
    "    print(seq_record1.description)\n",
    "    print(repr(seq_record1.seq))\n",
    "    print(\"Length %i\" % len(seq_record1))\n",
    "    print(seq_record1.annotations[\"keywords\"])\n",
    "    print(\"\\n\")\n",
    "\n",
    "    with ExPASy.get_sprot_raw(text2) as handle:\n",
    "        seq_record2 = SeqIO.read(handle, \"swiss\")\n",
    "    print(seq_record2.id)\n",
    "    print(seq_record2.name)\n",
    "    print(seq_record2.description)\n",
    "    print(repr(seq_record2.seq))\n",
    "    print(\"Length %i\" % len(seq_record2))\n",
    "    print(seq_record2.annotations[\"keywords\"])\n",
    "    print(\"\\n\")\n",
    "    \n",
    "    seq1=seq_record1.seq\n",
    "\n",
    "    seq2=seq_record2.seq\n",
    "\n",
    "    print(\"--------------------------------------------------------PROTEIN ALIGNMENT-------------------------------------------------------------\\n\")\n",
    "\n",
    "    for a in pairwise2.align.localxx(seq1, seq2):\n",
    "       \n",
    "        print(format_alignment(*a))\n",
    "        \n",
    "    return(text1,text2)\n",
    "\n",
    "def ramachadran():\n",
    "    \n",
    "    global text1\n",
    "       \n",
    "    text1=nombreID1.get()\n",
    "      \n",
    "    confProDy(auto_show=False)\n",
    "    confProDy(auto_secondary=True)\n",
    "    p38 = parsePDB(text1,compressed=False)\n",
    "    chain = p38['A']\n",
    "    Phi = []; Psi = []; c = []\n",
    "    for res in chain.iterResidues():\n",
    "        try:\n",
    "            phi = calcPhi(res)\n",
    "            psi = calcPsi(res)\n",
    "        except:\n",
    "            continue\n",
    "        else:\n",
    "            Phi.append(phi)\n",
    "            Psi.append(psi)\n",
    "            if res.getResname() == 'GLY':\n",
    "                c.append('black')\n",
    "            else:\n",
    "                secstr = res.getSecstrs()[0]\n",
    "                if secstr == 'H':\n",
    "                    c.append('red')\n",
    "                elif secstr == 'G':\n",
    "                    c.append('darkred')\n",
    "                elif secstr == 'E':\n",
    "                    c.append('blue')\n",
    "                else:\n",
    "                    c.append('grey')\n",
    "\n",
    "    scatter(Phi, Psi, c=c, s=10);\n",
    "    xlabel('Phi (degree)');\n",
    "    ylabel('Psi (degree)');\n",
    "    \n",
    "    return(text1)\n",
    "\n",
    "    \n",
    "        \n",
    "#\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\AQUÍ EMPIEZ EL LAUNCHER TKINTER/////////////////////////////////////////////////\n",
    "#\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\CORONALYZER 1.1.0////////////////////////////////////////////////////// \n",
    "root =tk.Tk()\n",
    "root.iconbitmap(\"icon.ico\")\n",
    "root.title(\"Coronalyzer1.0\")\n",
    "root.geometry(\"700x650\")\n",
    "root.resizable(False,False)\n",
    "\n",
    "text = tk.Text(root,bg=\"yellow\", height=1)\n",
    "text.pack()\n",
    "\n",
    "text.insert('4.0', '------------------------BIENVENIDO A LA DEMO DE CORONALYZER---------------------')\n",
    "\n",
    "menubar = tk.Menu(root)\n",
    "root.config(bg=\"#0059b3\", menu=menubar, width=300, height=300)\n",
    "imagen = PhotoImage(file=\"corona.gif\")\n",
    "my_label=tk.Label(root, image=imagen)\n",
    "my_label.place(x=0, y=0, relwidth=1, relheight=1,)\n",
    "filemenu=tk.Menu(menubar, tearoff=0)\n",
    "filemenu1=tk.Menu(menubar, tearoff=0)\n",
    "filemenu.add_command(label=\"Secuencia\",command=Analizar_Seq)\n",
    "filemenu.add_command(label=\"Comparar 3 Seq\",command=Analizar_Mult)\n",
    "filemenu.add_command(label=\"Comparar 2 Seq\",command=Analizar_Sim)\n",
    "filemenu.add_command(label=\"Mutaciones\",command=Analizar_Mut)\n",
    "filemenu.add_command(label=\"Protein Alignment\",command=protein_aln)\n",
    "filemenu.add_command(label=\"Ramachadran Plot\",command=ramachadran)\n",
    "filemenu.add_command(label=\"Exit\", command=root.quit)\n",
    "menubar.add_cascade(label=\"Analisis\", menu=filemenu)\n",
    "\n",
    "miFrame=tk.Frame(root)\n",
    "miFrame.pack()\n",
    "miFrame.config(bg=\"grey\")\n",
    "\n",
    "miNombre1= tk.StringVar()\n",
    "miNombre2= tk.StringVar()\n",
    "miNombre3= tk.StringVar()\n",
    "miNombre4=tk.StringVar()\n",
    "miNombre5= tk.StringVar()\n",
    "\n",
    "nombreID1=tk.Entry(miFrame, textvariable=miNombre1)\n",
    "nombreID1.grid(row=1, column=1, padx=10, pady=10)\n",
    "\n",
    "nombreID2=tk.Entry(miFrame, textvariable=miNombre2)\n",
    "nombreID2.grid(row=2, column=1, padx=10, pady=10)\n",
    "\n",
    "nombreID3=tk.Entry(miFrame, textvariable=miNombre3)\n",
    "nombreID3.grid(row=3, column=1, padx=10, pady=10)\n",
    "\n",
    "nombreID4=tk.Entry(miFrame, textvariable=miNombre4)\n",
    "nombreID4.grid(row=4, column=1, padx=10, pady=10)\n",
    "\n",
    "nombreID5=tk.Entry(miFrame, textvariable=miNombre5)\n",
    "nombreID5.grid(row=5, column=1, padx=10, pady=10)\n",
    "\n",
    "idLabel=tk.Label(miFrame, text=\"Archivo 1:\", width=10, height=1, borderwidth=3, relief=\"raised\")\n",
    "idLabel.grid(row=1, column=0, sticky=\"e\", padx=10, pady=10)\n",
    "\n",
    "idLabel=tk.Label(miFrame, text=\"Archivo 2:\", width=10, height=1, borderwidth=3, relief=\"raised\")\n",
    "idLabel.grid(row=2, column=0, sticky=\"e\", padx=10, pady=10)\n",
    "\n",
    "idLabel=tk.Label(miFrame, text=\"Archivo 3:\", width=10, height=1, borderwidth=3, relief=\"raised\")\n",
    "idLabel.grid(row=3, column=0, sticky=\"e\", padx=10, pady=10)\n",
    "\n",
    "idLabel=tk.Label(miFrame, text=\"Archivo Fasta:\", width=14, height=1, borderwidth=3, relief=\"raised\")\n",
    "idLabel.grid(row=4, column=0, sticky=\"e\", padx=10, pady=10)\n",
    "\n",
    "idLabel=tk.Label(miFrame, text=\"Archivo PDB:\", width=14, height=1, borderwidth=3, relief=\"raised\")\n",
    "idLabel.grid(row=5, column=0, sticky=\"e\", padx=10, pady=10)\n",
    "\n",
    "boton4=tk.Button(root, text=\"Limpiar todos los campos\", command=limpiarCampos).pack()\n",
    "boton3=tk.Button(root, text=\"Descargar Archivo Fasta\", command=descargar_fasta).pack()\n",
    "boton3=tk.Button(root, text=\"Descargar Archivo PDB\", command=descargar_PDB).pack()\n",
    "\n",
    "\n",
    "root.config(menu=menubar)\n",
    "\n",
    "root.mainloop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aecd6b2-1c46-4f78-8bd5-4425555322ac",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "#import Biopython\n",
    "import Bio as Bio\n",
    "#import our three important packages\n",
    "from Bio import SeqIO\n",
    "from Bio import AlignIO\n",
    "from Bio import Phylo\n",
    "#https://medium.com/geekculture/phylogenetic-trees-implement-in-python-3f9df96c0c32\n",
    "#https://biopython-cn.readthedocs.io/zh_CN/latest/en/chr13.html\n",
    "#import the sequences we will use. These are 16s sequences from GenBank\n",
    "#example: https://www.ncbi.nlm.nih.gov/nuccore/FJ039971.1?report=genbank \n",
    "t1 = Bio.SeqIO.read(\"rcsb_pdb_1P9B.fasta\", \"fasta\")\n",
    "t2 = Bio.SeqIO.read(\"rcsb_pdb_2V40.fasta\", \"fasta\")\n",
    "t3 = Bio.SeqIO.read(\"rcsb_pdb_5I34.fasta\", \"fasta\")\n",
    "t4 = Bio.SeqIO.read(\"rcsb_pdb_1DJ2.fasta\", \"fasta\")\n",
    "\n",
    "\n",
    "#rename each of the sequences \n",
    "#this step is not required, it will just make the tree easier to understand \n",
    "print(t1.description)\n",
    "print(t2.description)\n",
    "print(t3.description)\n",
    "\n",
    "\n",
    "t1.id = '1P9B'\n",
    "t2.id = '2V40'\n",
    "t3.id = '5I34'\n",
    "t4.id = '1DJ2'\n",
    "\n",
    "\n",
    "\n",
    "# Combine all of the individual sequences into a new file \n",
    "variantes = SeqIO.write([t1,t2,t3,t4], \"variantes.fasta\", \"fasta\")\n",
    "# Load the turtles sequences into MUSCLE \n",
    "#https://www.ebi.ac.uk/Tools/msa/muscle/\n",
    "# Upload the new alignment file to your folder or working directory \n",
    "# Open the alignment file as a MultipleSeqAlignment object \n",
    "with open(\"variantes.aln\",\"r\") as aln: \n",
    "    alignment = AlignIO.read(aln,\"clustal\")\n",
    "print(type(alignment))\n",
    "\n",
    "# Open and initiate the Distance Calculator using the Identity model \n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator \n",
    "calculator = DistanceCalculator('identity')\n",
    "# Write the Distance Matrix \n",
    "distance_matrix = calculator.get_distance(alignment)\n",
    "print(distance_matrix)\n",
    "\n",
    "# Open and initiate the Tree Constructor \n",
    "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor\n",
    "constructor = DistanceTreeConstructor(calculator)\n",
    "# Build the tree \n",
    "variant_tree = constructor.build_tree(alignment)\n",
    "variant_tree.rooted = True\n",
    "print(variant_tree)\n",
    "\n",
    "# Save the tree to a new file \n",
    "Phylo.write(variant_tree, \"Cov_tree.xml\", \"phyloxml\")\n",
    "\n",
    "# Import matplotlib and create a basic tree \n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "fig = Phylo.draw(variant_tree)\n",
    "\n",
    "# Make a better looking tree using the features of matplotlib \n",
    "\n",
    "fig = plt.figure(figsize=(13, 5), dpi=100) # create figure & set the size \n",
    "matplotlib.rc('font', size=12)              # fontsize of the leaf and node labels \n",
    "matplotlib.rc('xtick', labelsize=10)       # fontsize of the tick labels\n",
    "matplotlib.rc('ytick', labelsize=10)       # fontsize of the tick labels\n",
    "#turtle_tree.ladderize()\n",
    "axes = fig.add_subplot(1, 1, 1)\n",
    "Phylo.draw(variant_tree, axes=axes)\n",
    "fig.savefig(\"Covid_cladogram\")\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "69385296-e108-4177-b04d-5d494d3a3e27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "01492d9d5415407ea248a28635a45519",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hades\\anaconda3\\lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 23232.\n",
      "  warnings.warn(\n",
      "C:\\Users\\hades\\anaconda3\\lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 23386.\n",
      "  warnings.warn(\n",
      "C:\\Users\\hades\\anaconda3\\lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 23540.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chain ID: A\n",
      "chain ID: B\n",
      "chain ID: C\n",
      "chain ID: D\n",
      "chain ID: E\n",
      "chain ID: F\n",
      "chain ID: G\n",
      "chain ID: H\n",
      "chain ID: I\n",
      "chain ID: J\n",
      "chain ID: K\n",
      "chain ID: L\n",
      "chain ID: M\n",
      "chain ID: N\n",
      "chain ID: O\n",
      "chain ID: P\n",
      "chain ID: Q\n",
      "chain ID: R\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5faa8d0fc96f4fbdad9a67501289bd61",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "26183a7e2d734aa58d1a5c56e237b8b3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Tab(children=(Box(children=(Box(children=(Box(children=(Label(value='step'), IntSlider(value=1, min=-100)), la…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'''||||||||||||||||||||||||||||||||||||||||||||||||||VISUALIZACIÓN 3D DE LA PROTEÍNA DEL PDB||||||||||||||||||||||||||||||||||||||||||'''\n",
    "import nglview as nv\n",
    "from Bio.PDB import MMCIFParser\n",
    "\n",
    "def visualizar_proteina():\n",
    "    global structure\n",
    "    global text1\n",
    "       \n",
    "    text1=nombreID1.get()\n",
    "    \n",
    "    #Encuentra el archivo PDB\n",
    "    parser = MMCIFParser()\n",
    "    structure = parser.get_structure(text1, text1+'.cif')\n",
    "    structure\n",
    "    structure[0]\n",
    "    #identifica el numero d cadenas\n",
    "    for chain in structure[0]:\n",
    "        print(f'chain ID: {chain.id}')\n",
    "    nv.demo()\n",
    "    view = nv.show_biopython(structure)\n",
    "    view.render_image()\n",
    "    return (view, structure)\n",
    "\n",
    "visualizar_proteina()\n",
    "\n",
    "#GUI\n",
    "nv.show_biopython(structure, gui=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c0d38abb-37cc-4319-bc79-9b1a24e361bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def MUSCLE():\n",
    "    \n",
    "    global text1 \n",
    "    text1=nombreID1.get()\n",
    "    \n",
    "    in_file = \"C:/Users/hades/Coronalyzer/\"+text1\n",
    "    out_file = \"C:/Users/hades/Coronalyzer/aligned.fasta\"\n",
    "    muscle_exe = \"C:/Users/hades/Coronalyzer/muscle3.8.31_i86win32.exe\"\n",
    "    cline = MuscleCommandline(muscle_exe, input=in_file, phyiout=out_file)\n",
    "    stdout, stderr = cline(in_file)\n",
    "    align = AlignIO.read(out_file, \"phylip\")\n",
    "    print(align)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c51b699a-65ec-4c63-9ff8-cb959501c018",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6626a585",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6a966d6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
