{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a9773453-dc99-498a-b4cf-467562d318e1",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (Temp/ipykernel_10632/833136970.py, line 473)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"C:\\Users\\hades\\AppData\\Local\\Temp/ipykernel_10632/833136970.py\"\u001b[1;36m, line \u001b[1;32m473\u001b[0m\n\u001b[1;33m    print('Total proteins:', len(df))def conv(item):\u001b[0m\n\u001b[1;37m                                     ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import os\n",
    "import math\n",
    "#from tabulate import tabulate\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "pd.plotting.register_matplotlib_converters()\n",
    "#import seaborn as sns\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from tabulate import *\n",
    "pd.plotting.register_matplotlib_converters()\n",
    "import Bio\n",
    "import pylab\n",
    "import nglview as nv\n",
    "from Bio.PDB import PDBParser\n",
    "\n",
    "def  visualizar_proteina():\n",
    "    global structure\n",
    "    nombre=input(\"Inserte el codigo de la prote√≠na:\")\n",
    "    p = PDBParser(PERMISSIVE=1) \n",
    "    #find PDB file\n",
    "    parser = PDBParser()\n",
    "    structure = parser.get_structure('nombre', nombre+'.pdb')\n",
    "    structure\n",
    "    structure[0]\n",
    "    #identify number of chains\n",
    "    for chain in structure[0]:\n",
    "        print(f'chain ID: {chain.id}')\n",
    "    nv.demo()\n",
    "    view = nv.show_biopython(structure)\n",
    "    view.render_image()\n",
    "    nv.show_biopython(structure, gui=True)\n",
    "    return (view, structure)\n",
    "\n",
    "def Mutaciones():\n",
    "    \n",
    "    # This dictionary is used for reshaping the numpy array for each of the genes. \n",
    "    # For example, the gene ORF1ab numpy array will be reshaped with rows=115 and cols=115.\n",
    "    # The second element in the list represent 'n' number of dummy 'N' to be added at the end of\n",
    "    # each gene nucletode seq to make to compactible with the rows and cols of the array.\n",
    "    numpy_image_dict = {'gene=ORF1ab':[(115,115),7],\n",
    "                        'gene=S':[(62,62),22],\n",
    "                        'gene=ORF3a':[(28,30),12],\n",
    "                        'gene=E':[(15,16),12], \n",
    "                        'gene=M':[(26,27),33],\n",
    "                        'gene=ORF6':[(14,14),10],    \n",
    "                        'gene=ORF7a':[(19,20),14],\n",
    "                        'gene=ORF7b':[(12,12),12],\n",
    "                        'gene=ORF8':[(19,20),14],\n",
    "                        'gene=N':[(36,36),36],\n",
    "                        'gene=ORF10':[(11,11),4]}\n",
    "    \n",
    "    \n",
    "    # This dictionary has the codons for the amino acids from mRNA.\n",
    "    amacid_dict = {'UUU':('F','PHE','Phenylalanine'),'UUC':('F','PHE','Phenylalanine'),\n",
    "                   'UUA':('L','LEU','Leucine'),'UUG':('L','LEU','Leucine'),\n",
    "                   'UCU':('S','SER','Serine'),'UCC':('S','SER','Serine'),\n",
    "                   'UCA':('S','SER','Serine'),'UCG':('S','SER','Serine'), \n",
    "                   'UAU':('Y','TYR','Tyrosine'),'UAC':('Y','TYR','Tyrosine'),\n",
    "                   'UAA':('STOP'),'UAG':('STOP'),\n",
    "                   'UGU':('C','CYS','Cysteine'),'UGC':('C','CYS','Cysteine'),\n",
    "                   'UGA':('STOP'),'UGG':('W','TRP','Tryptophan'),\n",
    "                   'CUU':('L','LEU','Leucine'),'CUC':('L','LEU','Leucine'),\n",
    "                   'CUA':('L','LEU','Leucine'),'CUG':('L','LEU','Leucine'),\n",
    "                   'CCU':('P','PRO','Proline'),'CCC':('P','PRO','Proline'),\n",
    "                   'CCA':('P','PRO','Proline'),'CCG':('P','PRO','Proline'),\n",
    "                   'CAU':('H','HIS','Histidine'),'CAC':('H','HIS','Histidine'),\n",
    "                   'CAA':('Q','GLU','Glutamine'),'CAG':('Q','GLU','Glutamine'),\n",
    "                   'CGU':('R','ARG','Arginine'),'CGA':('R','ARG','Arginine'),\n",
    "                   'CGG':('R','ARG','Arginine'),'CGC':('R','ARG','Arginine'),\n",
    "                   'AUU':('I','ILE','Isoleucine'),'AUC':('I','ILE','Isoleucine'),\n",
    "                   'AUA':('I','ILE','Isoleucine'),'AUG':('M','MET','Methionine'),\n",
    "                   'ACU':('T','THR','Threonine'),'ACC':('T','THR','Threonine'),\n",
    "                   'ACA':('T','THR','Threonine'),'ACG':('T','THR','Threonine'),\n",
    "                   'AAU':('N','ASN','Asparagine'),'AAC':('N','ASN','Asparagine'),\n",
    "                   'AAA':('K','LYS','Lysine'),'AAG':('K','LYS','Lysine'),\n",
    "                   'AGU':('S','SER','Serine'),'AGC':('S','SER','Serine'),\n",
    "                   'AGG':('R','ARG','Arginine'),'AGA':('R','ARG','Arginine'),\n",
    "                   'GUU':('V','VAL','Valine'),'GUC':('V','VAL','Valine'),\n",
    "                   'GUA':('V','VAL','Valine'),'GUG':('V','VAL','Valine'),\n",
    "                   'GCU':('A','ALA','Alanine'),'GCC':('A','ALA','Alanine'),\n",
    "                   'GCA':('A','ALA','Alanine'),'GCG':('A','ALA','Alanine'),\n",
    "                   'GAU':('D','ASP','Aspartate'),'GAC':('D','ASP','Aspartate'),\n",
    "                   'GAA':('E','GLU','Glutamate'),'GAG':('E','GLU','Glutamate'),\n",
    "                   'GGU':('G','GLY','Glycine'),'GGC':('G','GLY','Glycine'),\n",
    "                   'GGA':('G','GLY','Glycine'),'GGG':('G','GLY','Glycine')\n",
    "    }\n",
    "    #This is the core module developed as part of the effort.\n",
    "    class dna:\n",
    "    \n",
    "        # Constructor method\n",
    "        def __init__(self,dna_seq):\n",
    "            dna_seq = dna_seq.upper() # Convert the nucleotide bases to Upper Case \n",
    "            for seq in dna_seq:\n",
    "                # Valid nucleotide bases. If not a valid sequence raise an Error\n",
    "                if seq not in ['A','T','G','C',' ','N']: \n",
    "                    error = 'Wrong DNA Sequence {}!!'.format(seq)\n",
    "                    raise ValueError(error)\n",
    "            # Remove all of the empty characters in the nucleotide sequence.\n",
    "            dna_seq = dna_seq.replace(' ','') \n",
    "            self.dir_3_5=dna_seq\n",
    "            self.dir_5_3=self.dir_5_3_strand()\n",
    "            self.mRna = None\n",
    "            self.amino_acid = None\n",
    "            self.num_array = None\n",
    "            self.nucl_len = len(dna_seq)\n",
    "    \n",
    "        def __repr__(self):\n",
    "            return \"DNA has {} nucleotide and they are {} :\".format(self.nucl_len,self.dir_3_5)\n",
    "    \n",
    "        def __eq__(self, other):\n",
    "            if other is None:\n",
    "                return False\n",
    "            return self.seq == other.seq\n",
    "    \n",
    "        #def replicate(self):\n",
    "        #    return\n",
    "    \n",
    "        def transcription(self):\n",
    "            # This is a method that imitates the transcription of a gene to mRNA for Protein transalation.\n",
    "            # This is mostly of the future use.\n",
    "            trans=''\n",
    "            for nuc in self.dir_5_3:\n",
    "                if nuc == 'A':\n",
    "                    trans += 'U'\n",
    "                if nuc == 'T':\n",
    "                    trans += 'A'\n",
    "                if nuc == 'C':\n",
    "                    trans += 'G'\n",
    "                if nuc == 'G':\n",
    "                    trans += 'C'\n",
    "                if nuc == 'N':\n",
    "                    trans += 'N'\n",
    "            self.mRna = trans\n",
    "            return self.mRna\n",
    "    \n",
    "        def translation(self):\n",
    "            # This is the method where the transcripted mRNA gets translated into Amino Acid. Each 3\n",
    "            # base in the mRNA codes for an amino acid.\n",
    "            begin = 'No'\n",
    "            ac = ''\n",
    "            for i in range(0,len(self.mRna)-3,3):\n",
    "                if self.mRna[i:3] == 'AUG':\n",
    "                    begin = 'Yes'\n",
    "                if self.mRna[i:3] in ('UAA','UAG','UGA'):\n",
    "                    being = 'No'\n",
    "                if begin == 'Yes':\n",
    "                    ac+= amacid_dict[self.mRna[i:3+i]][0]\n",
    "            self.amino_acid = ac\n",
    "            return self.amino_acid\n",
    "    \n",
    "        def dir_5_3_strand(self):\n",
    "            dir_5_3 = ''\n",
    "            # This is a method which reads the 3 - 5 prime sequence and creates the 5 - 3 prime sequence.\n",
    "            for nuc in self.dir_3_5:\n",
    "                if nuc == 'A':\n",
    "                    dir_5_3 += 'T'\n",
    "                if nuc == 'T':\n",
    "                    dir_5_3 += 'A'\n",
    "                if nuc == 'C':\n",
    "                    dir_5_3 += 'G'\n",
    "                if nuc == 'G':\n",
    "                    dir_5_3 += 'C'\n",
    "                if nuc == 'N':\n",
    "                    dir_5_3 += 'N'\n",
    "            return dir_5_3\n",
    "    \n",
    "        def numpfy(self):\n",
    "            # This method takes in a dna sequence and convert them into numpy array.\n",
    "            # Each of the nucleotide sequence is converted into one of the below numbers \n",
    "            # which then can be used in for analysis and comparison.\n",
    "            arr = ''\n",
    "            for i in self.dir_3_5:\n",
    "                if i == 'A':\n",
    "                    arr += '0 '\n",
    "                if i == 'T':\n",
    "                    arr += '255 '\n",
    "                if i == 'C':\n",
    "                    arr += '100 '\n",
    "                if i == 'G':\n",
    "                    arr += '200 '\n",
    "                if i == 'N':\n",
    "                    arr += '75 '   \n",
    "            arr_np = np.fromstring(arr,dtype=np.uint8,sep=' ')        \n",
    "            self.num_array = arr_np\n",
    "            return self.num_array\n",
    "        \n",
    "        \n",
    "        \n",
    "    def read_dna_seq(file_name):\n",
    "        # This method reads the dna sequence from the file downloaded from NCBI and crates a python dictionary.\n",
    "        fil = open(file_name,'r')\n",
    "        fil_list = fil.readlines()\n",
    "        fil.close\n",
    "        \n",
    "        genome = {}\n",
    "        gene_name = ''\n",
    "        protein_name = ''\n",
    "        gene_seq = ''\n",
    "        for i in fil_list:\n",
    "            if i[0] == '>':\n",
    "                # Reads each line from the file and creates a dictionary with the following information for each\n",
    "                # gene. {<'gene_name-1'>:[<protein_name>,nucleotide sequence],\n",
    "                #        <'gene_name-2'>:[<protein_name>,nucleotide sequence],\n",
    "                #        <'gene_name-2'>:[<protein_name>,nucleotide sequence]}\n",
    "                if list(genome.keys()) != []:\n",
    "                    gene_seq = gene_seq.replace('\\n','')\n",
    "                    genome[gene_name].append(gene_seq)\n",
    "                gene_seq = ''\n",
    "                g_st = i.find('[gene=')\n",
    "                g_end = i[g_st:].find(']')\n",
    "                p_st = i.find('[protein=')\n",
    "                p_end = i[p_st:].find(']') \n",
    "    \n",
    "                if g_st > 0 and g_end > 0:\n",
    "                    gene_name = i[g_st+1:g_st+g_end]\n",
    "                    genome[gene_name] = []\n",
    "                \n",
    "                if p_st > 0 and p_end > 0:\n",
    "                    protein_name = i[p_st+1:p_st+p_end]\n",
    "                    genome[gene_name].append(protein_name)\n",
    "            else:\n",
    "                gene_seq += i\n",
    "        gene_seq = gene_seq.replace('\\n','')\n",
    "        genome[gene_name].append(gene_seq)    \n",
    "        return genome\n",
    "    \n",
    "    def gene_mod(genome):\n",
    "        # This method modifies each of the sequence with dummy nucleotide 'N' so that for the shape of the numpy array.\n",
    "        genome_keys = list(genome.keys())\n",
    "        for k in genome_keys:\n",
    "            if len(numpy_image_dict[k]) > 1:\n",
    "                N = numpy_image_dict[k][1]\n",
    "                seq = add_N(N,genome[k][1])\n",
    "                genome[k][1] = seq\n",
    "        return genome\n",
    "    \n",
    "    def add_N(n,seq):\n",
    "        # This method is called from gene_mod() method, for creating dummy nucleotide 'N'.\n",
    "        for i in range(0,n):\n",
    "            seq += 'N'\n",
    "        return seq\n",
    "    \n",
    "    nombre1=input(\"ingrese nombre del archivo (con el .txt por favor): \")\n",
    "    nombre2=input(\"ingrese nombre del archivo (con el .txt por favor): \")\n",
    "    # Read the dna sequence file-1 previously downloaded from NCBI.\n",
    "    dict_seq_1 = read_dna_seq(nombre1)\n",
    "    # Modify the sequence with dummy 'N' nucleotide.\n",
    "    dict_seq_1 = gene_mod(dict_seq_1)\n",
    "    \n",
    "    # Read the dna sequence file-2 previously downloaded from NCBI.\n",
    "    dict_seq_2 = read_dna_seq(nombre2)\n",
    "    # Modify the sequence with dummy 'N' nucleotide.\n",
    "    dict_seq_2 = gene_mod(dict_seq_2)\n",
    "    \n",
    "    # Create matplotlib subplots for each gene. \n",
    "    f,ax = plt.subplots(nrows=11,ncols=3,figsize=(25,30))\n",
    "    gene_name = list(numpy_image_dict.keys())\n",
    "    row = 0\n",
    "    col = 0\n",
    "    mut_dict={}\n",
    "    for i in gene_name:\n",
    "        G = i[5:]\n",
    "        # Loop thru each gene in the Cornona Virus nucleotide sequence.\n",
    "        gene_us = dna(dict_seq_1['gene='+G][1])\n",
    "        # Invoke the transcription method of the class dna \n",
    "        gene_us.transcription()\n",
    "        # Invoke the mothod that converts the gene sequence into a numpy array.\n",
    "        numpfy_usa = gene_us.numpfy()\n",
    "        # Reshape the numpy array with a predeifned shape from the numpy_image_dict dictionary.\n",
    "        numpfy_usa = numpfy_usa.reshape(numpy_image_dict['gene='+G][0])\n",
    "        # sub-plot the numpy array with matplotlib pcolor method.\n",
    "        ax[row][col].pcolor(numpfy_usa)\n",
    "        ax[row][col].set_title(G+' Gene - USA')\n",
    "        col+=1\n",
    "        gene_china = dna(dict_seq_2['gene='+G][1])\n",
    "        # Invoke the transcription method of the class dna \n",
    "        gene_china.transcription()\n",
    "        # Invoke the mothod that converts the gene sequence into a numpy array.\n",
    "        numpfy_china = gene_china.numpfy()\n",
    "        # Reshape the numpy array with a predeifned shape from the numpy_image_dict dictionary.\n",
    "        numpfy_china = numpfy_china.reshape(numpy_image_dict['gene='+G][0])\n",
    "        # sub-plot the numpy array with matplotlib pcolor method.\n",
    "        ax[row][col].pcolor(numpfy_china)\n",
    "        ax[row][col].set_title(G+' Gene - CHINA')\n",
    "        col+=1\n",
    "    \n",
    "        # To find the gene mutation subtract the numpy array from base sequence with the newer sequence. Here the \n",
    "        # the Chinese sequence is the base sequence and the USA sequence is a newer sequence.\n",
    "        mut = numpfy_china - numpfy_usa\n",
    "        if mut.any():\n",
    "            # Here we are looking for a non zero value in the mutated numpy array (result of the subtracting the 2 numpy arrays).\n",
    "            # Presence of non-zero value means that there is difference between the 2 numpy arrays and the gene has \n",
    "            # mutataions. If there are mutations in the gene create a python dictionary \"mut_dict\" with details as below.\n",
    "            # {'<Gene_Name-1>': [[<value_of_base_seq>, <value_of_newer_seq>, <value_in_mutated_numpy>, (x_value,y_value)]], '<Gene_Name-2>': [[<value_of_base_seq>, <value_of_newer_seq>, <value_in_mutated_numpy>, (x_value,y_value)]]}\n",
    "            mut_nec = np.nonzero(mut)\n",
    "            x=mut_nec[0]\n",
    "            y=mut_nec[1]\n",
    "            l=0\n",
    "            mut_dict[G]=[]\n",
    "            for i in x:\n",
    "                us_base = numpfy_usa[i][y[l]]\n",
    "                ch_base = numpfy_china[i][y[l]]\n",
    "                mut_base = mut[i][y[l]]\n",
    "                info_list = [ch_base,us_base,mut_base,(i,y[l])]\n",
    "                mut_dict[G].append(info_list)\n",
    "                print(\"Mutated DNA Base {} in China and Base {} in USA at position {} For the Gene {}\".format(ch_base,us_base,(i,y[l]),G))\n",
    "                l+= 1\n",
    "        # Giving a title to the matplotlib subplot\n",
    "        ax[row][col].pcolor(mut)\n",
    "        ax[row][col].set_title(G+' Gene - Mutataion')\n",
    "        row+= 1\n",
    "        col=0\n",
    "    \n",
    "    f.tight_layout()\n",
    "    # Saving the matplotlib subplot as a jpg.\n",
    "    f.savefig('Sars_Cov-2_Gene_Mutation.jpg')\n",
    "\n",
    "def AnalisiSeq():\n",
    "    \n",
    "    nombre3=input(\"ingrese nombre del archivo (con el .fna por favor): \")\n",
    "    coronavirus = open(nombre3, \"r\")\n",
    "    print('HEADER:',coronavirus.readline())\n",
    "    coronavirus = coronavirus.readlines()\n",
    "    COVID_seq = ''\n",
    "    for line in coronavirus:\n",
    "        line = line.strip()\n",
    "        COVID_seq += line\n",
    "    print(COVID_seq[0:1000])\n",
    "    \n",
    "    def basic_properties(DNAseq):\n",
    "        total_base = len(DNAseq)\n",
    "        num_Adenine = DNAseq.count('A')\n",
    "        num_Guanine = DNAseq.count('G')\n",
    "        num_Thymine = DNAseq.count('T')\n",
    "        num_Cytosine = DNAseq.count('C')\n",
    "        \n",
    "        if total_base != num_Adenine + num_Guanine + num_Thymine + num_Cytosine:\n",
    "            print('Something is not right')\n",
    "        else : pass\n",
    "        \n",
    "        A_percent = num_Adenine / total_base\n",
    "        G_percent = num_Guanine / total_base\n",
    "        T_percent = num_Thymine / total_base\n",
    "        C_percent = num_Cytosine / total_base\n",
    "        \n",
    "        #visualization\n",
    "        x = np.arange(4)\n",
    "        bases = ['Adenine', 'Guanine', 'Thymine' ,'Cytosine']\n",
    "        values = [num_Adenine, num_Guanine, num_Thymine, num_Cytosine]\n",
    "        plt.bar(x,values)\n",
    "        plt.xticks(x, bases)\n",
    "        plt.show()\n",
    "        table = [['total base',total_base,'Percentage',str('100%')],\n",
    "                 ['Adenine:',num_Adenine, 'Percentage:',str(round(A_percent*100,2))+'%'],\n",
    "                ['Guanine:',num_Guanine, 'Percentage:',str(round(G_percent*100,2))+'%'],\n",
    "                 ['Thynime:',num_Thymine, 'Percentage:',str(round(T_percent*100,2))+'%'],\n",
    "                 ['Cytosine:',num_Cytosine, 'Percentage:',str(round(C_percent*100,2))+'%']]\n",
    "        print(tabulate(table))\n",
    "        print('GC content:', round((((num_Guanine + num_Cytosine) / total_base)*100),2),'%')\n",
    "        \n",
    "    print(basic_properties(COVID_seq))\n",
    "    \n",
    "    def transcription(DNAseq):\n",
    "        mRNAseq = DNAseq.replace('T','U')\n",
    "        print(len(mRNAseq))\n",
    "        return mRNAseq\n",
    "    COVID_mRNA = transcription(COVID_seq)\n",
    "    print(\"--------------------------mRNA sequence------------------------------\")\n",
    "    print(COVID_mRNA[0:1000])\n",
    "    print(\"---------------------------Protein Sequence-----------------------------\")\n",
    "    def translate(seq):\n",
    "        table = {\n",
    "            'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "            'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "            'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "            'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n",
    "            'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "            'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "            'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "            'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "            'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "            'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "            'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "            'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "            'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "            'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "            'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "            'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',}\n",
    "        protein =\"\"\n",
    "        for i in range(0, len(seq), 3):\n",
    "            codon = seq[i:i + 3]\n",
    "            if codon in table:\n",
    "                protein+= table[codon]\n",
    "        return protein\n",
    "    \n",
    "    protein_seq = translate(COVID_seq)\n",
    "    print(protein_seq[0:1000])\n",
    "    \n",
    "    def visualization(protein_seq):\n",
    "        # composition of Amino Acids\n",
    "        plt.figure(figsize=(35,50))\n",
    "        x = np.arange(22)\n",
    "        AA = ['Arginine', 'Histidie','Lysine',\n",
    "              'Aspartic Acid','Glutamic Acid',\n",
    "              'Serine','Threonine','Asparagine','Glutamine',\n",
    "              'Cysteine','Selenocysteine','Glycine','Prolien',\n",
    "              'Alanine','Valine','Isoleucine','Leucine',\n",
    "              'Methionine','Phenylalanine','Tyrosine','Tryptophan',\n",
    "              'Stop Codon']\n",
    "        values = [protein_seq.count('R'),protein_seq.count('H'),protein_seq.count('K'),\n",
    "                  protein_seq.count('D'),protein_seq.count('E'),\n",
    "                  protein_seq.count('S'),protein_seq.count('T'),protein_seq.count('N'),protein_seq.count('Q'),\n",
    "                  protein_seq.count('C'),protein_seq.count('U'),protein_seq.count('G'),protein_seq.count('P'),\n",
    "                  protein_seq.count('A'),protein_seq.count('V'),protein_seq.count('I'),protein_seq.count('L'),\n",
    "                  protein_seq.count('M'),protein_seq.count('F'),protein_seq.count('Y'),protein_seq.count('W'),\n",
    "                  protein_seq.count('_')]\n",
    "        plt.subplot(2,2,1)\n",
    "        plt.rc('font',size = 20)\n",
    "        plt.barh(AA,values,height=0.6)\n",
    "        plt.title('AA in general')\n",
    "        \n",
    "        # visualization by groups as shown in figure\n",
    "        x = np.arange(4)\n",
    "        Electric =  protein_seq.count('R')+protein_seq.count('H')+protein_seq.count('K')+protein_seq.count('D')+protein_seq.count('E')\n",
    "        Uncharged = protein_seq.count('S')+protein_seq.count('T')+protein_seq.count('N')+protein_seq.count('Q')\n",
    "        Special =   protein_seq.count('C')+protein_seq.count('U')+protein_seq.count('G')+protein_seq.count('P')\n",
    "        Hydrophobic = protein_seq.count('A')+protein_seq.count('V')+protein_seq.count('I')+protein_seq.count('L')+protein_seq.count('M')+protein_seq.count('F')+protein_seq.count('Y')+protein_seq.count('W')\n",
    "        \n",
    "        plt.subplot(2,2,2)\n",
    "        types = ['Elecrically charged','Polar uncharged',\n",
    "                 'Special case', 'Hydrophobic Side Chain']\n",
    "        values = [Electric,Uncharged,Special,Hydrophobic]\n",
    "        plt.barh(types, values, height = 0.6)\n",
    "        plt.title('AA in groups')\n",
    "        \n",
    "        # Visualization of positive and negative amino acid in Hydrophobic side chain group\n",
    "        plt.subplot(2,2,3)\n",
    "        x = np.arange(2)\n",
    "        positive = protein_seq.count('R')+protein_seq.count('H')+protein_seq.count('K')\n",
    "        negative = protein_seq.count('D')+protein_seq.count('E')\n",
    "        types = ['Positive','Negative']\n",
    "        values = [positive, negative]\n",
    "        plt.barh(types,values, height=0.6)\n",
    "        plt.title('Charge diff in Electrically charged side chain')\n",
    "        plt.show()\n",
    "        \n",
    "        # Visualization abundance of Amino Acid by relative frequenceies of amino aicd residue in secondary structure\n",
    "        alpha_helix = protein_seq.count('A')+protein_seq.count('C')+protein_seq.count('L')+protein_seq.count('M')+protein_seq.count('E')+protein_seq.count('Q')+protein_seq.count('H')+protein_seq.count('K')\n",
    "        beta_sheet =protein_seq.count('V')+protein_seq.count('I')+protein_seq.count('F')+protein_seq.count('Y')+protein_seq.count('W')+protein_seq.count('T')\n",
    "        turn = protein_seq.count('G')+protein_seq.count('S')+protein_seq.count('D')+protein_seq.count('N')+protein_seq.count('P')\n",
    "        x = np.arange(3)\n",
    "        plt.subplot(2,2,4)\n",
    "        types = ['Alpha helix','beta sheet','Turn']\n",
    "        values = [alpha_helix, beta_sheet, turn]\n",
    "        plt.barh(types,values, height=0.6)\n",
    "        plt.title('AA residues in secondary structure')\n",
    "        plt.show()\n",
    "   \n",
    "    print(visualization(protein_seq))\n",
    "    print(\"---------------------------Protein Splicing-----------------------------\")\n",
    "    def proteinseq_splic(protein_sequence):\n",
    "        protein_group = protein_sequence.split('_')\n",
    "        return protein_group\n",
    "    protein_list = proteinseq_splic(protein_seq)\n",
    "    print(proteinseq_splic(protein_seq)[0:100])\n",
    "    df = pd.DataFrame(Proteins)\n",
    "df.describe()\n",
    "print('Total proteins:', len(df))def conv(item):\n",
    "    return len(item)def to_str(item):\n",
    "    return str(item)df['sequence_str'] = df[0].apply(to_str)\n",
    "df['length'] = df[0].apply(conv)\n",
    "df.rename(columns={0: \"sequence\"}, inplace=True)\n",
    "df.head()# Take only longer than 20\n",
    "functional_proteins = df.loc[df['length'] >= 20]\n",
    "print('Total functional proteins:', len(functional_proteins))\n",
    "functional_proteins.describe()\n",
    "    print('Number of proteins:',len(protein_list))\n",
    "    def Functional_protein(protein_list):\n",
    "        funct_group = []\n",
    "        for protein in protein_list:\n",
    "            if len(protein) > 20:\n",
    "                funct_group.append(protein)\n",
    "        return funct_group\n",
    "    function_protein = Functional_protein(protein_list)\n",
    "    print(\"---------------------------Functional Protein-----------------------------\")\n",
    "    print(function_protein[0:52])\n",
    "    print('Number of Functional proteins:',len(function_protein))\n",
    "    \n",
    "    dna_ORF = []\n",
    "    \n",
    "    def rf(seq):\n",
    "        #getting all reading frames\n",
    "        i=0\n",
    "        while i+2<len(seq):\n",
    "            if seq[i:i+3]=='ATG':\n",
    "                j=i+3\n",
    "                while j+2<len(seq):\n",
    "                    if seq[j:j+3] in ['TGA', 'TAA', 'TAG']:\n",
    "                        dna_ORF.append(seq[i:j+3])\n",
    "                    j=j+3\n",
    "            i=i+3\n",
    "        return dna_ORF\n",
    "    \n",
    "    def RC(seq):\n",
    "        #reverse compliment\n",
    "        complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "        bases = list(seq) \n",
    "        bases = reversed([complement.get(base,base) for base in bases])\n",
    "        bases = ''.join(bases)\n",
    "        return bases\n",
    "       \n",
    "    def ORF(seq):\n",
    "        #open reading frame\n",
    "        #num=0\n",
    "        seqRC=RC(seq)\n",
    "        for i in [0,1,2]:\n",
    "            rf(seq[i:])\n",
    "            rf(seqRC[i:])\n",
    "            \n",
    "    RC(COVID_seq)[0:1000]\n",
    "    \n",
    "    print(\"-----------------------------DNA ORF---------------------------------\")\n",
    "    ORF(COVID_seq)\n",
    "    print(dna_ORF[0:10])\n",
    "    print(len(dna_ORF))\n",
    "    \n",
    "    # converting dnaORF to proteinORF\n",
    "    candidate_pro_ORF = []\n",
    "    def dna_pro_ORF(dnaORF):\n",
    "        for ORF in dnaORF:\n",
    "            candidate_pro_ORF.append(translate(ORF))\n",
    "        return candidate_pro_ORF\n",
    "\n",
    "    print(dna_pro_ORF(dna_ORF)[0:10])\n",
    "    \n",
    "    valid_group = []\n",
    "    def valid_ORF(protein_seq):\n",
    "        for protein in protein_seq:\n",
    "            if protein.count('_') == 1:\n",
    "                valid_group.append(protein)\n",
    "        return valid_group\n",
    "    \n",
    "    print(valid_group[0:10])\n",
    "    len(valid_group)   \n",
    "    \n",
    "    functional_ORF = Functional_protein(valid_group)\n",
    "    print(functional_ORF[0:20])\n",
    "    print('number of functional protein:',len(functional_ORF))\n",
    "    \n",
    "    sort_functional_ORF = list(sorted(functional_ORF, key = len, reverse=True))\n",
    "    largest_FP = sort_functional_ORF[0]\n",
    "    print(largest_FP[0:1000])\n",
    "    print(len(largest_FP))\n",
    "    visualization(largest_FP)\n",
    "    \n",
    "    second_FP = sort_functional_ORF[1]\n",
    "    print(second_FP[0:1000])\n",
    "    print('protein size:',len(second_FP))\n",
    "    visualization(second_FP)\n",
    "    \n",
    "    third_FP = sort_functional_ORF[2]\n",
    "    print(third_FP[0:1000])\n",
    "    print('protein size:',len(third_FP))\n",
    "    visualization(third_FP)\n",
    "    \n",
    "def Comparar():\n",
    "    nombre4=input(\"ingrese nombre del archivo (con el .fasta por favor): \")\n",
    "    nombre5=input(\"ingrese nombre del archivo (con el .fasta por favor): \")\n",
    "    nombre6=input(\"ingrese nombre del archivo (con el .fasta por favor): \")\n",
    "    SARS = SeqIO.read(nombre4, 'fasta')\n",
    "    MERS = SeqIO.read(nombre5, 'fasta')\n",
    "    COV2 = SeqIO.read(nombre6, 'fasta')\n",
    "    \n",
    "    #Squiggle wont work in Kaggle kernels, you need to run this command on terminal.\n",
    "    \n",
    "    \n",
    "    # Alignments using pairwise2 alghoritm\n",
    "    \n",
    "    SARS_COV = pairwise2.align.globalxx(SARS.seq, COV2.seq, one_alignment_only=True, score_only=True)\n",
    "    print('SARS/COV Similarity (%):', SARS_COV / len(SARS.seq) * 100)\n",
    "    \n",
    "    MERS_COV = pairwise2.align.globalxx(MERS.seq, COV2.seq, one_alignment_only=True, score_only=True)\n",
    "    print('MERS/COV Similarity (%):', MERS_COV / len(MERS.seq) * 100)\n",
    "    \n",
    "    MERS_SARS = pairwise2.align.globalxx(MERS.seq, SARS.seq, one_alignment_only=True, score_only=True)\n",
    "    print('MERS/SARS Similarity (%):', MERS_SARS / len(SARS.seq) * 100)\n",
    "    \n",
    "    # Plot the data\n",
    "    X = ['SARS/COV2', 'MERS/COV2', 'MERS/SARS']\n",
    "    Y = [SARS_COV/ len(SARS.seq) * 100, MERS_COV/ len(MERS.seq)*100, MERS_SARS/len(SARS.seq)*100]\n",
    "    plt.title('Sequence identity (%)')\n",
    "    plt.bar(X,Y,color=(0.2, 0.4, 0.6, 0.6))\n",
    "\n",
    "\n",
    "\n",
    "def descargar_fasta():\n",
    "    archivo=input(\"Ingrese el nombre del archivo.fasta:\")\n",
    "\n",
    "    Entrez.email = \"mingyan24@126.com\"\n",
    "    hd1 = Entrez.efetch(db=\"nucleotide\",id=[archivo],rettype='fasta')\n",
    "    seq = SeqIO.read(hd1,'fasta')\n",
    "    fw = open(archivo,'w')\n",
    "    SeqIO.write(seq,fw,'fasta')\n",
    "    fw.close()\n",
    "    os.getcwd()\n",
    "    \n",
    "def Menu():\n",
    " \n",
    "    correcto=False\n",
    "    num=0\n",
    "    while(not correcto):\n",
    "        try:\n",
    "            num = int(input(\"Elige metodo de an√°lisis: \"))\n",
    "            correcto=True\n",
    "        except ValueError:\n",
    "            print('Error, vuelve a elegir')\n",
    "     \n",
    "    return num\n",
    " \n",
    "salir = False\n",
    "opcion = 0\n",
    " \n",
    "while not salir:\n",
    " \n",
    "    print (\"1. Mutaciones entre variantes\")\n",
    "    print (\"2. An√°lisis de Secuencia Covid\")\n",
    "    print (\"3. Comparacion secuencia variantes\")\n",
    "    print (\"4. Visualizar proteina\")\n",
    "    print (\"5. Salir\")\n",
    "     \n",
    "    print (\"Elige una opcion\")\n",
    " \n",
    "    opcion = Menu()\n",
    " \n",
    "    if opcion == 1:\n",
    "        Mutaciones()\n",
    "    elif opcion == 2:\n",
    "        AnalisiSeq()\n",
    "    elif opcion == 3:\n",
    "        Comparar()\n",
    "    elif opcion== 4: \n",
    "        visualizar_proteina()\n",
    "    elif opcion == 5:\n",
    "        salir = True\n",
    "    else:\n",
    "        print (\"Elige una opci√≥n\")\n",
    " \n",
    "print (\"Fin\")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "430fa234-fae5-44d6-8dee-d59fa0e9c040",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a44a9ba3f5ec4e0fa2fb97ce8dcd2c17",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Introduzca el codigo de la proteina para descargar:6vxx\n",
      "Structure exists: 'PDB\\6vxx.cif' \n"
     ]
    }
   ],
   "source": [
    "import Bio\n",
    "import pylab\n",
    "import pandas as pd\n",
    "import nglview as nv\n",
    "from Bio.PDB import MMCIFParser\n",
    "from Bio.PDB import PDBList\n",
    "\n",
    "archivo=input(\"Introduzca el codigo de la proteina para descargar:\")\n",
    "pdbl=PDBList()\n",
    "pdbl1=[archivo]\n",
    "pdbl.download_pdb_files(pdbl1, pdir='PDB', file_format='mmCif')\n",
    "\n",
    "def  visualizar_proteina():\n",
    "    global structure\n",
    "    nombre=input(\"Inserte el codigo de la prote√≠na:\")\n",
    "    p = PDBParser(PERMISSIVE=1) \n",
    "    #find PDB file\n",
    "    parser = PDBParser()\n",
    "    structure = parser.get_structure('nombre', nombre+'.pdb')\n",
    "    structure\n",
    "    structure[0]\n",
    "    #identify number of chains\n",
    "    for chain in structure[0]:\n",
    "        print(f'chain ID: {chain.id}')\n",
    "    nv.demo()\n",
    "    view = nv.show_biopython(structure)\n",
    "    view.render_image()\n",
    "    return (view, structure)\n",
    "\n",
    "visualizar_proteina()\n",
    "\n",
    "#GUI\n",
    "nv.show_biopython(structure, gui=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe6c3ef3-864f-46cd-8a3e-24268df08278",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
